qiskit/result/counts.py

```diff
@@ -80,9 +81,13 @@ def __init__(self, data, time_taken=None, creg_sizes=None, memory_slots=None):
                     self.hex_raw = {hex(key): value for key, value in self.int_raw.items()}
                 else:
                     if not creg_sizes and not memory_slots:
+                        # bit strings with no leading 0b or dit strings
                         self.hex_raw = None
                         self.int_raw = None
-                        bin_data = data
+                        if all(isinstance(k, str) for k in data):
+                            bin_data = data
+                        else:
+                            raise ValueError("All keys must be of the same type")
```

Comment:
I think lint will complain about `ValueError` not being in the docstring since it's a new exception type being raised here.

Reply:
Right. I didn't run it yet. But, I recall that a linting tool (pylint?) has complained about this in the past. The linting tool is wrong. I'd like to put the exception comment on the class definition. I'll look into modifying the pylint rule. It would be pretty hard to get a linter to reliably tell when to put it in `Raises` and when not to.

Reply:
> I'm on the fence about that I'm not sure we really need to document this as a potential change that requires user input as part of the release.

I'm on the fence too. The only reason not to document it that I can think of is to avoid cluttering the release notes with trivia There are good arguments for documenting it. I suppose someone might be intentionally or inadvertently using `Counts` with mixed keys in some way that never raises an error down the line.

Reply:
I wrote a long argument against the missing-raises-doc thing in a comment in an already open issue: https://github.com/Qiskit/qiskit-terra/issues/1179#issuecomment-1234664993

Reply:
I did a fair amount of digging into `missing-raises-doc` including what people do in other languages. Dropped some of that in #1179. If it's easier to get this merged without adding the pylint exception, that is, documenting the `raise`, I'm ok with that.

Reply:
I changed this: 1) Changed `ValueError` -> `TypeError`. 2) Documented this in the docstring. This should resolve this conversation.